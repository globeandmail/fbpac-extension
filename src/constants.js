// Constants
export const ToggleType = {
  ADS: "ads",
  RATER: "rater",
  STORIES: "stories"
};

export const RatingType = {
  POLITICAL: "political",
  NORMAL: "normal"
};

export const MOST_RECENT_STORIES_UPDATE = 1;
