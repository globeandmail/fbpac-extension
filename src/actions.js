import { sendAds, adForRequest } from "utils.js";
import { RatingType } from "constants.js";

// Action Types
export const ACCEPT_TERMS = "accept_terms";
export const TOGGLE_TAB = "toggle_tab";
export const NEW_ADS = "new_ads";
export const NEW_RATINGS = "new_ratings";
export const UPDATE_AD = "update_ad";
export const UPDATE_RATING = "update_rating";
export const SET_LANGUAGE = "set_language";
export const SET_COUNTRY = "set_country";
export const SAY_THANKS = "say_thanks";

// Actions
export const setLanguage = language => ({ type: SET_LANGUAGE, language });
export const setCountry = country => ({ type: SET_COUNTRY, country });
export const acceptTerms = () => ({ type: ACCEPT_TERMS });
export const toggle = (newActiveTab, oldActiveTab) => ({
  type: TOGGLE_TAB,
  value: newActiveTab,
  oldTab: oldActiveTab
});

export const newAds = ads => ({
  type: NEW_ADS,
  value: ads
});

export const newRatings = ratings => ({
  type: NEW_RATINGS,
  value: ratings
});

export const updateAd = (id, rating) => ({
  type: UPDATE_AD,
  id: id,
  value: rating
});

export const updateRating = (id, rating) => ({
  type: UPDATE_RATING,
  id: id,
  value: rating
});

export const sayThanks = ratings_count => ({
  type: SAY_THANKS,
  ratings_count
});

export const rateAd = (ad, rating, update) => {
  return (dispatch, getState) => {
    let body = {
      ...adForRequest(ad),
      political: rating === RatingType.POLITICAL
    };
    dispatch(update(ad.id, rating));
    dispatch(sayThanks(getState().ratings_count || 0));
    let cb = () => ({});
    return sendAds([body], getState().language).then(cb, cb);
  };
};
