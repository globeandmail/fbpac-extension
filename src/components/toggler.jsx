import React from "react";
import { connect } from "react-redux";
import { toggle } from "actions.js";
import { withI18n } from "i18n.js";
import { countUnratedRatings } from "utils.js";
import { ToggleType } from "constants.js";
import { Ads, UnratedRatings } from "components/ads.jsx";
import { Stories } from "components/stories.jsx";
// Controls which section of tabs to show, defaults to the user's ads
export const Toggle = withI18n(
  ({ getMessage, type, message, active, amount, dispatch }) => (
    <div
      className={"toggle" + (active === type ? " active" : "")}
      onClick={() => dispatch(toggle(type, active))}
    >
      {getMessage(message)}
      {amount ? <b>{100 > amount ? amount : "100+"}</b> : ""}
    </div>
  )
);

// Our Main container.
const TogglerUnconnected = ({ ads, ratings, active, language }) => (
  <div id="toggler">
    <div id="tabs">
      <Toggle
        amount={countUnratedRatings(ratings)}
        active={active}
        message="rate_ads"
        type={ToggleType.RATER}
      />
      <Toggle
        amount={countUnratedRatings(ads)}
        active={active}
        message="see_ads"
        type={ToggleType.ADS}
      />
      {(language && (language.language === "en-US" || language.language === "en-CA")) || true ? (
        <Toggle
          amount={1}
          active={active}
          message="stories"
          type={ToggleType.STORIES}
        />
      ) : null}
    </div>
    <div id="container">
      {active === ToggleType.ADS ? (
        <Ads ads={ads} />
      ) : active === ToggleType.STORIES ? (
        <Stories />
      ) : (
        <UnratedRatings ratings={ratings} />
      )}
    </div>
  </div>
);

export const Toggler = connect(state => state)(TogglerUnconnected);
