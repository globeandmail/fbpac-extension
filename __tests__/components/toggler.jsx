import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import { ToggleType } from "../../src/constants.js";
import { Toggle, Toggler } from "../../src/components/toggler.jsx";
import { toggle } from "../../src/actions.js";

import configureMockStore from "redux-mock-store";

const mockStore = configureMockStore();
describe("toggle", () => {
  it("should render a toggle", () => {
    const politicalStore = mockStore({
      active: ToggleType.RATER,
      language: {}
    });
    const political = mount(
      <Provider store={politicalStore}>
        <Toggle type={ToggleType.RATER} amount={1000} />
      </Provider>
    );
    expect(political).toMatchSnapshot();
    const normalStore = mockStore({
      active: ToggleType.ADS,
      amount: 0,
      language: {}
    });
    const normal = mount(
      <Provider store={normalStore}>
        <Toggle type={ToggleType.ADS} amount={0} />
      </Provider>
    );
    expect(normal).toMatchSnapshot();
    normal
      .find(".toggle")
      .at(0)
      .simulate("click");
    expect(normalStore.getActions()[0]).toEqual(toggle(ToggleType.ADS));
  });
});

describe("toggler", () => {
  it("should render a toggler", () => {
    const politicalStore = mockStore({
      active: ToggleType.RATER,
      language: {},
      ads: [],
      ratings: []
    });
    const political = mount(
      <Provider store={politicalStore}>
        <Toggler />
      </Provider>
    );
    expect(political).toMatchSnapshot();
    const normalStore = mockStore({
      active: ToggleType.ADS,
      language: {},
      ads: [],
      ratings: []
    });
    const normal = mount(
      <Provider store={normalStore}>
        <Toggler />
      </Provider>
    );
    expect(normal).toMatchSnapshot();
  });
});
