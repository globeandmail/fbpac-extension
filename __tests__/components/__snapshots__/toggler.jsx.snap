// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toggle should render a toggle 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Component)
    amount={1000}
    type="rater"
  >
    <Component
      amount={1000}
      dispatch={[Function]}
      getMessage={[Function]}
      language={Object {}}
      type="rater"
    >
      <div
        className="toggle"
        onClick={[Function]}
      >
        TK(undefined)
        <b>
          100+
        </b>
      </div>
    </Component>
  </Connect(Component)>
</Provider>
`;

exports[`toggle should render a toggle 2`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Component)
    amount={0}
    type="ads"
  >
    <Component
      amount={0}
      dispatch={[Function]}
      getMessage={[Function]}
      language={Object {}}
      type="ads"
    >
      <div
        className="toggle"
        onClick={[Function]}
      >
        TK(undefined)
      </div>
    </Component>
  </Connect(Component)>
</Provider>
`;

exports[`toggler should render a toggler 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(TogglerUnconnected)>
    <TogglerUnconnected
      active="rater"
      ads={Array []}
      dispatch={[Function]}
      language={Object {}}
      ratings={Array []}
    >
      <div
        id="toggler"
      >
        <div
          id="tabs"
        >
          <Connect(Component)
            active="rater"
            amount={0}
            message="rate_ads"
            type="rater"
          >
            <Component
              active="rater"
              amount={0}
              dispatch={[Function]}
              getMessage={[Function]}
              language={Object {}}
              message="rate_ads"
              type="rater"
            >
              <div
                className="toggle active"
                onClick={[Function]}
              >
                Ads I'm Seeing
              </div>
            </Component>
          </Connect(Component)>
          <Connect(Component)
            active="rater"
            amount={0}
            message="see_ads"
            type="ads"
          >
            <Component
              active="rater"
              amount={0}
              dispatch={[Function]}
              getMessage={[Function]}
              language={Object {}}
              message="see_ads"
              type="ads"
            >
              <div
                className="toggle"
                onClick={[Function]}
              >
                Ads Others are Seeing
              </div>
            </Component>
          </Connect(Component)>
        </div>
        <div
          id="container"
        >
          <Connect(Ratings)
            ratings={Array []}
          >
            <Ratings
              onRatingClick={[Function]}
              ratings={Array []}
            >
              <div
                id="ratings"
              />
            </Ratings>
          </Connect(Ratings)>
        </div>
      </div>
    </TogglerUnconnected>
  </Connect(TogglerUnconnected)>
</Provider>
`;

exports[`toggler should render a toggler 2`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(TogglerUnconnected)>
    <TogglerUnconnected
      active="ads"
      ads={Array []}
      dispatch={[Function]}
      language={Object {}}
      ratings={Array []}
    >
      <div
        id="toggler"
      >
        <div
          id="tabs"
        >
          <Connect(Component)
            active="ads"
            amount={0}
            message="rate_ads"
            type="rater"
          >
            <Component
              active="ads"
              amount={0}
              dispatch={[Function]}
              getMessage={[Function]}
              language={Object {}}
              message="rate_ads"
              type="rater"
            >
              <div
                className="toggle"
                onClick={[Function]}
              >
                Ads I'm Seeing
              </div>
            </Component>
          </Connect(Component)>
          <Connect(Component)
            active="ads"
            amount={0}
            message="see_ads"
            type="ads"
          >
            <Component
              active="ads"
              amount={0}
              dispatch={[Function]}
              getMessage={[Function]}
              language={Object {}}
              message="see_ads"
              type="ads"
            >
              <div
                className="toggle active"
                onClick={[Function]}
              >
                Ads Others are Seeing
              </div>
            </Component>
          </Connect(Component)>
        </div>
        <div
          id="container"
        >
          <Connect(AdsUnconnected)
            ads={Array []}
          >
            <AdsUnconnected
              ads={Array []}
              onAdClick={[Function]}
            >
              <div
                id="ads"
              />
            </AdsUnconnected>
          </Connect(AdsUnconnected)>
        </div>
      </div>
    </TogglerUnconnected>
  </Connect(TogglerUnconnected)>
</Provider>
`;
